### PR TITLE
ruby-build: force use of homebrew openssl

### PR DESCRIPTION
**Pull request by CamJN**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

I know that https://github.com/Homebrew/homebrew-core/pull/46637 exists, however I'd like to suggest that this is a better approach to the problem. 

As is ruby-build will vendor a specific openssl version, per ruby installed, at install time, and then never update it. 

Some of the consequences of ruby-build's approach are:

- When openssl 1.1.1e or w/e the next security update is for openssl is called, none of your rbenv installed rubies will be updated.
- You have n extra copies of openssl lying around.
- In order to use self signed certs with your ruby's openssl, you need to add them to every single openssl and regenerate the cert bundle.

Some consequences of making this change:

- Some rubies won't build
- The ruby-build maintainers might get annoyed

Some alternatives I considered:

- just revert commit https://github.com/rbenv/ruby-build/commit/a0dcf52834ed76ee6baba3a03ffbd7d9399c2c79 as a patch in the formula, but that seems fragile
- do nothing, but that seems irresponsible
- add a scary warning to the ruby-build caveats warning users about the issue, this seems like it could be a reasonable compromise

### CODE REVIEW

**Comment by igas** [comment]
> When openssl 1.1.1e or w/e the next security update is for openssl is called, none of your rbenv installed rubies will be updated.

How does your approach solves it? I don't remember rbenv recompiling every installed ruby on `brew upgrade openssl`.

Homebrew tries not to dictate how packages work. We can not maintain every package and go against how project works. I believe this should be addressed in `ruby-build`.

cc @mislav @gfguthrie

**Comment by igas** [comment]
I'm not sure if it was proposed/discussed, but I would do `brew install openssl@1.1` in `ruby-build` if brew is installed for rubies that support it. We do not ship 1.0 anymore due to EOL, if `ruby-build` ok with shipping it, for this case current behaviour will work for older rubies.

**Comment by CamJN** [comment]
@igas I'm pretty sure ruby is dynamically linked to openssl so when you replace openssl in homebrew all the rubies will pick up the new version. IMO that's the most important reason for a package manager.

**Comment by igas** [comment]
@CamJN I believe shipping working software somewhere in the important reasons too.

**Comment by CamJN** [comment]
@igas oh absolutely, and it is just my opinion after all, but I think that security vulnerabilities are a pretty severe form of broken software, and not being able to use rubies that depend on such broken software isn't too bad, given the alternative. 

There's also the principle of least surprise to consider. If I install a package with homebrew, I expect the dependencies to be managed by homebrew and upgraded when I `brew upgrade`.

**Comment by igas** [comment]
To be clear, I do agree that security is important. I'm just not sure that proposed approach is the best we can do. Let's wait for `ruby-build` maintainers to respond and see what can we do about it. I'm sure everyone have best intention.

> There's also the principle of least surprise to consider. If I install a package with homebrew, I expect the dependencies to be managed by homebrew and upgraded when I brew upgrade.

I would not be so sure, I do not expect homebrew to update rubies just because I installed `ruby-build`, or if I install `ruby` for homebrew too update gems.

**Comment by mislav** [comment]
> Some of the consequences of ruby-build's approach are:
> 
> * When openssl 1.1.1e or w/e the next security update is for openssl is called, none of your rbenv installed rubies will be updated.
> * You have n extra copies of openssl lying around.
> * In order to use self signed certs with your ruby's openssl, you need to add them to every single openssl and regenerate the cert bundle.

These are all valid points, and if you are concerned about them you are welcome to ``export RUBY_CONFIGURE_OPTS=--with-openssl-dir="$(brew --prefix openssl@1.1)"`` in your shell initialization file and have all your rubies link to Homebrew's openssl.

It would be ideal if this was the default in ruby-build, and it was up until recently, but it caused too many problems for our users, and all those problems would simply come back if this PR was merged.

> Some consequences of making this change:
> 
> * Some rubies won't build

Like @igas points out, this is a pretty big dealbreaker. It matters little that Ruby installs are "secure" (re: openssl upgrades) if they are non-functional to begin with.

I'm 👎 on this PR, obviously, since it's too opinionated and does nothing else other than reverting a recent change to ruby-build.

**Comment by gfguthrie** [comment]
Not a maintainer – just started the mess, and brought it to its current conclusion.

This PR re-creates all of the problems we just went through (and IMO exacerbates them through a blurring of separation of concerns). This will at minimum break installing ruby for every version prior to 2.5. "But they're not officially maintained anymore." Correct, but `ruby-build` still supports installing those versions, and the public still uses them (not acknowledging this was my original conceit).

The reasoning behind undoing everything was the pileup of support issues from the first go-round. We've put the onus on the people who want to use Homebrew-installed `openssl` to specifically use it when they install ruby (via `--with-openssl-dir`), so that `ruby-build` can continue to work for the vast majority that just want something that works.

**Comment by igas** [comment]
@CamJN do you feel like adding caveat with `RUBY_CONFIGURE_OPTS` would be a good solution?

**Comment by CamJN** [comment]
@igas that should be fine as long as it makes it clear that if you do it, it'll mess up building EOL rubies, but if you don't you'll never get fixes for openssl vulnerabilities.

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
> I'm 👎 on this PR, obviously, since it's too opinionated and does nothing else other than revert a recent change to ruby-build (that was done after much deliberation from our end).

With this in mind I'd rather defer to upstream maintainers and say this PR should just be a `RUBY_CONFIGURE_OPTS` caveats change.

**Comment by chdiza** [comment]
> If I install a package with homebrew, I expect the dependencies to be managed by homebrew and upgraded when I brew upgrade

Depends on what the package is.  If I install python with homebrew, I don't expect homebrew to manage the dependencies of everything that I `pip install`.  That's pip's job.  Likewise, I would not have ever expected a brewed ruby-build to have the deps of the individual rubies it builds be managed by homebrew.

File: Formula/ruby-build.rb
Comment (mislav) [inline Formula/ruby-build.rb]: I would rephrase this as:

> By default, ruby-build installs a copy of OpenSSL for each individual Ruby version. Such instance of OpenSSL is never updated.

File: Formula/ruby-build.rb
Comment (mislav) [inline Formula/ruby-build.rb]: I would be more specific here:

> To have all Ruby versions link to Homebrew's OpenSSL, which can be upgraded over time, you can place this in your shell initialization file:

File: Formula/ruby-build.rb
Comment (mislav) [inline Formula/ruby-build.rb]: Should we specify OpenSSL 1.1 here? This is to guard ourselves from future potential `openssl@1.2` which might be incompatible with current Ruby versions
```suggestion
      export RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1)"
```

File: Formula/ruby-build.rb
Comment (mislav) [inline Formula/ruby-build.rb]: Again, let's be more specific here:

> Note that this might interfere with building some versions of Ruby, such as those older than Ruby 2.4.

**Review by mislav — CHANGES_REQUESTED** [review]
This caveat is only applicable on macOS. Can we ensure that this caveat is only printed on macOS and not on Linux?

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**
